identity-apps-core/apps/recovery-portal/src/main/webapp/password-recovery-otp.jsp

```diff
@@ -0,0 +1,315 @@
+<%@ page import="org.wso2.carbon.identity.mgt.endpoint.util.client.model.recovery.v2.RecoveryResponse" %>
+<%@ page import="org.wso2.carbon.identity.mgt.endpoint.util.client.model.recovery.v2.ResendRequest" %>
+<%@ page import="org.wso2.carbon.identity.mgt.endpoint.util.client.model.recovery.v2.ResendResponse" %>
+<%@ page import="org.wso2.carbon.identity.mgt.endpoint.util.IdentityManagementEndpointConstants" %>
+
+<%-- Include tenant context --%>
+<%@ include file="tenant-resolve.jsp"%>
+
+<%! 
+    /**
+     * RecoveryStage represents the four steps of recovery in the 
+     * PasswordRecoveryAPI V2.
+     */
+    public enum RecoveryStage {
+
+        INITIATE("INITIATE"),
+        RESEND("RESEND"),
+        CONFIRM("CONFIRM"),
+        RESET("RESET");
+
+        private final String value;
+
+        RecoveryStage(String value) {
+            
+            this.value = value;
+        }
+
+        /**
+         * Returns tye string value of the enum.
+         */
+        public String getValue() {
+
+            return value;
+        }
+
+        /**
+         * Override the toStirng method of the object class.
+         */
+        @Override
+        public String toString() {
+
+            return value;
+        }
+
+        /**
+         * Compares enum with string based on it's value.
+         * @return boolean whether the passed string equals the value of the enum.
+         */
+        public boolean equalsValue(String otherValue) {
+
+            return this.value.equals(otherValue);
+        }
+    }
+%>
+<%! 
+    /**
+    * This generates a random number to provide a sample number for a invalid username
+    * to avoid letting external entities learning of the existing usernames. The same 
+    * random number string will be generated for the same username each time.
+    */
+    public static String getRandomNumberString(int len, String seed) {
+
+        StringBuilder sb = new StringBuilder(len);
+        Random random = new Random(seed.hashCode());
+
+        for (int i = 0; i < len; i++) {
+            Integer numAtIndex = random.nextInt(10);
+            sb.append(numAtIndex.toString());
+        }
+        return sb.toString();
+    }
+%>
+<%!
+    /**
+     * This redirects the flow to the error page with the provided error message.
+     */
+    public void redirectToErrorPageWithMessage(HttpServletRequest request, 
+        HttpServletResponse response, String errorMsg) throws ServletException, IOException {
+        
+        request.setAttribute("error", true);
+        request.setAttribute("errorMsg", errorMsg);
+        request.getRequestDispatcher("error.jsp").forward(request, response);
+    }
+%>
+<%
+    final String USERNAME_CLAIM_URI_FOR_TENANT = "http://wso2.org/claims/username";
+    final RecoveryApiV2 recoveryApiV2 = new RecoveryApiV2();
+    final String username = IdentityManagementEndpointUtil.getStringValue(request.getAttribute("username"));
```

Comment:
Could u check whether we need to encode request.getAttribute method calls as well?

Reply:
request.getAttribute method is intended to be used on server side. Hence it is not needed to encode the value.